oe_paragraphs.install

```diff
@@ -0,0 +1,172 @@
+      $entity->save();
+    }
+
+    // New form modes.
+    $form_modes = $storage->listAll('core.entity_form_mode.');
+    foreach ($form_modes as $mode) {
+      $values = $storage->read($mode);
+      $form_mode = EntityFormMode::create($values);
+      $form_mode->save();
+    }
+
+    // Update form displays.
+    $form_displays = $storage->listAll('core.entity_form_display.');
+    foreach ($form_displays as $display) {
+      $values = $storage->read($display);
+      $form_display = EntityFormDisplay::load($values['id']);
+      if ($form_display) {
+        foreach ($values as $key => $value) {
+          $form_display->set($key, $value);
+        }
+      }
+      else {
+        $form_display = EntityFormDisplay::create($values);
+      }
+
+      $form_display->save();
+    }
+
+    // Update view displays.
+    $view_displays = $storage->listAll('core.entity_view_display.');
+    foreach ($view_displays as $display) {
+      $values = $storage->read($display);
+      $view_display = EntityViewDisplay::load($values['id']);
+      foreach ($values as $key => $value) {
+        $view_display->set($key, $value);
+      }
+
+      $view_display->save();
+    }
+  }
+
+  // Return if we have nothing to process.
+  if ($sandbox['total'] === 0) {
+    \Drupal::messenger()->addMessage($sandbox['current'] . ' paragraphs processed.');
+    $sandbox['#finished'] = 1;
+
+    // Cleanup the fields after copy.
+    foreach ($sandbox['variant_fields'] as $bundle => $field) {
+      FieldStorageConfig::loadByName('paragraph', $field)->delete();
+    }
+
+    return;
+  }
+
+  // Get the current slice of paragraph ids.
+  $pids = array_slice($paragraph_ids, $sandbox['current'], $sandbox['paragraphs_per_batch']);
+  $entity_storage = \Drupal::entityTypeManager()->getStorage('paragraph');
+
+  foreach ($pids as $pid) {
+    $paragraph = $entity_storage->load($pid);
+    $bundle = $paragraph->bundle();
+    $value = $paragraph->get($sandbox['variant_fields'][$bundle])->getValue();
+
+    // Check for values that needs conversion.
+    if (strpos($value[0]['value'], 'list_item_') !== FALSE) {
+      $value[0]['value'] = substr($value[0]['value'], strlen('list_item_'));
+    }
+
+    // Take out list block item values as they are not variants but layouts.
+    if ($sandbox['variant_fields'][$bundle] === 'field_oe_list_item_block_variant') {
+      // Set the layout filed with the value.
+      $paragraph->set('field_oe_list_item_block_layout', $value);
+    }
+    else {
+      // Set the new field with the old value.
+      $paragraph->set('field_oe_paragraphs_variant', $value);
+    }
+
+    // Set the old field to NULL.
+    $paragraph->set($sandbox['variant_fields'][$bundle], NULL);
+    $paragraph->save();
+
+    $sandbox['current']++;
+  }
+
+  \Drupal::messenger()->addMessage($sandbox['current'] . ' paragraphs processed.');
+  $sandbox['#finished'] = ($sandbox['current'] / $sandbox['total']);
```

Comment:
Maybe you can do like `system_update_8001()`:
```
$sandbox['#finished'] = empty($sandbox['total']) ? 1 : ($sandbox['current'] / $sandbox['total']);
```
So you can skip the check on `$sandbox['total']` and the code duplication above.

Reply:
Yes, but in this case we need to run the deletion before the fields then return before the rest of the code is being executed.